### PR TITLE
Stepless 5: don't read flow steps for flow resumption

### PR DESCRIPTION
**Pull request by rowanseymour**
Or save rule_uuid/rule_value on them since that's no longer read anywhere

### CODE REVIEW

File: temba/flows/models.py [hunk 1]
```diff
@@ -2020,18 +2022,16 @@ def add_step(self, run, node, msgs=(), exit_uuid=None, is_start=False, previous_
 
         if previous_step:
             previous_step.left_on = arrived_on
-            previous_step.rule_uuid = exit_uuid
             previous_step.next_uuid = node.uuid
-            previous_step.save(update_fields=('left_on', 'rule_uuid', 'next_uuid'))
```
Comment (rowanseymour) [inline temba/flows/models.py]: somewhat unrelated - just don't need this any more

File: temba/flows/models.py [hunk 2]
```diff
@@ -830,21 +830,24 @@ def find_and_handle(cls, msg, started_flows=None, voice_response=None,
...
-            flow = step.run.flow
+        for run in FlowRun.get_active_for_contact(msg.contact):
+            flow = run.flow
             flow.ensure_current_version()
-            destination = Flow.get_node(flow, step.step_uuid, step.step_type)
+
+            last_step = run.path[-1]
+            destination = Flow.get_node(flow, last_step[FlowRun.PATH_NODE_UUID], FlowStep.TYPE_RULE_SET)
+
+            step_obj = run.steps.order_by('id').last()
```
Comment (rowanseymour) [inline temba/flows/models.py]: still get the step object IF it exists and update it as per usual

File: temba/flows/models.py [hunk 3]
```diff
@@ -856,10 +859,10 @@ def find_and_handle(cls, msg, started_flows=None, voice_response=None,
                 if flowserver_trial:
                     trial_result = None
 
-                    if msg and msg.id:
-                        trial_result = trial.end_resume(flowserver_trial, msg_in=msg)
-                    elif expired_child_run:
+                    if expired_child_run:
```
Comment (rowanseymour) [inline temba/flows/models.py]: we sometimes have both a message and an expired run so this needs to come first

File: temba/flows/models.py [hunk 4]
```diff
@@ -3561,42 +3574,46 @@ def continue_parent_flow_run(cls, run, trigger_send=True, continue_parent=True):
...
+            msg.id = 0
+            msg.text = ''
+            msg.org = run.org
+            msg.contact = run.contact
+
+        if run.exit_type == FlowRun.EXIT_TYPE_EXPIRED:
+            allow_trial = True
+            expired_child_run = run
+        else:
+            allow_trial = False
```
Comment (rowanseymour) [inline temba/flows/models.py]: As much as a I hate adding another to `Flow.find_and_handle`, we don't want it to do flow server trials inside flow server trials, which can happen when this method is being used in the subflow case.

Comment (ericnewcomer) [reply temba/flows/models.py]: man its a good thing `find_and_handle` isn't long for this world anyways

**Pull request merged**